### ISSUE TITLE
optional runtype isn't really optional on records

### ISSUE THREAD

**Issue by pabra** [hoeck/simple-runtypes#18]
Your first example in the [readme](https://github.com/hoeck/simple-runtypes#example) actually isn't correct (anymore?).
The `email` key isn't really optional in a way that it can be absent too. See [this codesandbox](https://codesandbox.io/s/new-framework-g4ifo?file=/src/index.ts).

For the runtype check it seems to be fine, if the `email` key is absent. But in the resulting typescript type (`ReturnType<typeof userRuntype>`) the key isn't marked optional.

I would expect this Type:
```typescript
type User = {
    id: number;
    name: string;
    email?: string | undefined;
}
```
but actually it's this:
```typescript
type User = {
    id: number;
    name: string;
    email: string | undefined;
}
```

**Comment by pabra** [comment]
This seems to work (also [codesandbox](https://codesandbox.io/s/pensive-bogdan-17b7j?file=/src/index.ts)):
```typescript
const userRuntype = st.intersection(
  st.record({
    id: st.integer(),
    name: st.string()
  }),
  st.partial(
    st.record({
      email: st.string()
    })
  )
);
```
If that's the "official" solution for optional keys on records, I think the readme should be updated.

**Comment by hoeck** [comment]
Oh, you found a blind spot. The reason for this behavior is that `optional` only works on the value, and `?` is an annotation thats required on the `record` (aka the interface) key.

`optional` cannot do that because its not limited to records, for example you could write `const myArray = st.array(st.optional(st.string()))` to get a runtype of `(string | undefined)[]`.

Not sure how to solve this. One possible solution would be to rename the current optional to sth like `orUndefined` (or completely remove it as its just a shortcut for `st.union(st.undefined(), <whatever>)` and then to build a new `optional` to only work on records.

The solution you present looks like a usable workaround but I wouldn't want add that to the readme. It does not look very readable, especially when compared to the interface it represents: a simple `T & {email?: string}` ends up as `T & Partial<{email: string}>`.

So, any ideas on how to implement `st.optional(T)` such that turns 
`record({o: st.optional(T), r: S})` into `{o?: T, r: S}`
while ensuring that:
* it can only be used within `record` (to avoid misusing it, maybe its okay to implement this as a runtime check)
* it does not result in a hyper-complicated `record`-type (because for me, runtypes form the basis of my application and want them to be as plain and simple as possible)

**Comment by pabra** [comment]
Another workaround could be to explicitly type it:
```typescript
const userRuntypeRecord: {
  id: st.Runtype<number>;
  name: st.Runtype<string>;
  email?: st.Runtype<string | undefined>;
} = {
  id: st.integer(),
  name: st.string(),
  email: st.optional(st.string())
};
const userRuntype = st.record(userRuntypeRecord);
```
or cast it:
```typescript
const userRuntype = st.record({
  id: st.integer(),
  name: st.string(),
  email: st.optional(st.string())
} as {
  id: st.Runtype<number>;
  name: st.Runtype<string>;
  email?: st.Runtype<string | undefined>;
});
```

**Comment by hoeck** [comment]
fixed in 4477d00

**Comment by hoeck** [comment]
fixed in 4477d005cf2652a8b2881530477c9feaee61d104

**Comment by hoeck** [comment]
fixed in 4477d005cf2652a8b2881530477c9feaee61d104
